intrusive_collections_examples/src/basic_datastructures/red_black_tree.rs

```diff
@@ -0,0 +1,359 @@
+///     object: T,
+///     color: RedBlackTreeNodeColor,
+///     parent: Option<*const OwningRedBlackTreeNode<T>>
+///     left_child: Option<*const OwningRedBlackTreeNode<T>>,
+///     right_child: Option<*const OwningRedBlackTreeNode<T>>
+/// }
+/// ```
+/// which has storage for the type that we'd like to store.
+///
+/// An intrusive data structure flips this on it's head; each instance of `T`
+/// owns a single node of the red-black tree.  So, given only a reference to a
+/// `RedBlackTreeNode` node, how do you get back the `T` that it's a part of?
+/// The trick is that the compiler knows how `T` is laid out in memory.  Given
+/// a pointer to a `T`, we can calculate how many bytes there are to reach where
+/// the `RedBlackTreeNode` node is laid out.  Here is a compressed example of
+/// what we're talking about:
+/// ```rust
+/// #[derive(Debug)]
+/// pub enum RedBlackTreeNodeColor {
+///     Red,
+///     Black,
+/// }
+///
+/// impl Default for RedBlackTreeNodeColor {
+///     fn default() -> Self {
+///         RedBlackTreeNodeColor::Black
+///     }
+/// }
+///
+/// #[derive(Debug, Default)]
+/// pub struct RedBlackTreeNode {
+///     color: RedBlackTreeNodeColor,
+///     parent: Option<*const RedBlackTreeNode>
+///     left_child: Option<*const RedBlackTreeNode>,
+///     right_child: Option<*const RedBlackTreeNode>
+/// }
+///
+/// #[derive(Debug, Default)]
+/// pub struct Thing {
+///     field: Vec<u8>,
+///     next: RedBlackTreeNode,
+/// }
+///
+/// fn main() {
+///     let thing = Thing::default();
+///     let ptr1: *const Thing = &thing;
+///
+///     // We could use any field we wanted, we're using `parent` just as an
+///     // example here.
+///     let ptr2: *const RedBlackTreeNode = &(thing.parent);
+///     let offset = unsafe { (ptr2 as *const u8).offset_from(ptr1 as *const u8) };
+///     println!(
+///         "The 'parent' field is offset from the start of 'thing' by {:?} bytes",
+///         offset
+///     );
+///     assert!(offset >= 0);
+///     let ptr3: *const Thing = unsafe { ((ptr2 as *const u8).sub(offset as usize)) as *const Thing };
+///     assert_eq!(ptr1, ptr3);
+/// }
+/// ```
+///
+/// As you can see, we're doing a fair amount of casting, and there is some
+/// `unsafe` code within the example.  This is a cost of intrusive collections,
+/// and one that you should be aware of when implementing them.  Which brings up
+/// the talk about safety...
+///
+/// ## Safety, or, how to shoot yourself in the foot
+///
+/// Intrusive collections are generally unsafe to work with.  They require
+/// working with raw pointers instead of references, and this can't easily be
+/// changed.  For a more in-depth explanation of why this is so, read the
+/// [`README.md`](../README.md) file.
+#[derive(Debug, Clone, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct RedBlackTreeNode<T>
+where
+    T: Debug + Clone + Default + PartialEq + Eq + PartialOrd + Ord + Hash,
```

Comment:
1. It's generally a bad idea to put bounds on type definitions, it makes them harder to use for little benefit.
2. `Ord` implies the other comparison traits, so putting `T: Ord` also implies `T: PartialEq + Eq + PartialOrd`

Reply:
Yeah, I agree that it's generally a bad idea, but there's a reason for it here; if the containing object wants to derive those methods as suggested by the [Rust API Guidelines](https://rust-lang.github.io/api-guidelines/interoperability.html#c-common-traits), then `T` has to implement them as well.  Otherwise anyone that wants to use it would have to manually implement the traits...

I'll change to just using `Ord` though, that will definitely be more compact.

Reply:
Yes, and the derive macro will handle that case. You don't need to restrict `RedBlackTreeNode` to *just* the types that are `Debug + Clone + Ord + Hash`.

For example, [`HashMap`](https://doc.rust-lang.org/std/collections/struct.HashMap.html) doesn't put *any* trait bounds on the type itself, only on the impl blocks that need them. i.e. `HashMap::insert` requires `Hash + Eq` but `HashMap::new` doesn't. This allows you to use `HashMap` more flexibly. (For example, the keys don't need `Hash + Eq` at all, so long as you provide a way to hash/compare it from outside using the nightly `raw_entry` API).

There is a trend in the `std` towards moving trait bounds as close to the actual use site as possible, and I tend to agree with this trend. Of course, there are reasons to put trait bounds in the type (for example, when that trait bound informs the layout of the type), but these cases are rare and I don't think they apply here.

Reply:
Good point!  OK, I'm logging another issue, and will think about how to move the trait bounds around.

That said, I think I will keep `Debug` as a bounds on the keys, not because they are required in general, but because these examples are intended as a learning tool, and `Debug` helps with debugging code.  Hmmmm... that actually gives me an idea, I'll try to figure out how to pretty-print the debug output of trees.